File: Backend/src/presentation/controllers/TypesRolesController.ts
import { Request, Response } from "express";
import { ParamsDictionary } from "express-serve-static-core";
import { ParsedQs } from "qs";
import { ITypesRolesController } from "./interfaces/ITypesRolesController";
import { ITypesRolesServices } from "../../services/interfaces/ITypesRolesServices";
import { typesRolesServices } from "../../services/typeRolesServices";

import { typesRoles } from "../../domain/Entities/typesRoles";
import { IdSchema, TypeNameSchema, createSchema, updateSchema } from "./Schema/TypesRolesSchema";

class TypesRolesController implements ITypesRolesController{
    private typesRolesServices: ITypesRolesServices;
    constructor(){
        this.typesRolesServices = new typesRolesServices();
        this.GetByTypeName = this.GetByTypeName.bind(this);
        this.getAll = this.getAll.bind(this);
        this.create = this.create.bind(this);
        this.getId = this.getId.bind(this);
        this.update = this.update.bind(this);
    }
    async GetByTypeName(req: Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>, res: Response<any, Record<string, any>>): Promise<Response<any, Record<string, any>>> {
        const resultSchema = await TypeNameSchema.safeParseAsync(req.body);
        if(resultSchema.success){
            const { typeName } = resultSchema.data;
            const typeRole = this.typesRolesServices.GetTypeRoleByTypeName(typeName);
            return res.json(typeRole);
        }
        return res.json(resultSchema.error);
    }
    async create(req: Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>, res: Response<any, Record<string, any>>): Promise<Response<any, Record<string, any>>> {
        const resultSchema = await createSchema.safeParseAsync(req.body);
        if(resultSchema.success){
            const { typeName } = resultSchema.data;
            const typeRole = new typesRoles();
            typeRole.id = 0;
            typeRole.typeName = typeName;
            const result = await this.typesRolesServices.Save(typeRole);
            return res.json(result);
        }
        return res.json(resultSchema.error);
    }
    async update(req: Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>, res: Response<any, Record<string, any>>): Promise<Response<any, Record<string, any>>> {
        const resultSchema = await updateSchema.safeParseAsync(req.body);
        if(resultSchema.success){
            const { id, typeName } = resultSchema.data;
            const typeRole = new typesRoles();
            typeRole.id = id;
            typeRole.typeName = typeName;
            const result = await this.typesRolesServices.Save(typeRole);
            return res.json(result);
        }
        return res.json(resultSchema.error);
    }
    async getAll(req: Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>, res: Response<any, Record<string, any>>): Promise<Response<any, Record<string, any>>> {
        const list =  await this.typesRolesServices.GetAll();
        if(list.length !== 0 ){
            return res.json(list);
        }
        return res.json("Nada Encontrado!");
    }
    delete(req: Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>, res: Response<any, Record<string, any>>): Promise<Response<any, Record<string, any>>> {
        throw new Error("Method not implemented.");
    }
    async getId(req: Request<ParamsDictionary, any, any, ParsedQs, Record<string, any>>, res: Response<any, Record<string, any>>): Promise<Response<any, Record<string, any>>> {
        const resultSchema = await IdSchema.safeParseAsync(req.body);
        if(resultSchema.success){
            const { id } = resultSchema.data;
            const typeRole = this.typesRolesServices.GetbyId(id);
            if(typeRole !== undefined){
                return res.json(typeRole);
            }
            return res.json("Nada encontrado");
        }
        return res.json(resultSchema.error);
    }

}
export { TypesRolesController }

File: Back/src/presentation/Controllers/TypesRolesController.py
from flask import Blueprint, request, jsonify
import json
from services.typesRolesServices import typesRolesServices
from domain.Entities.typesRoles import typesRoles
typesRolesController = Blueprint('TypesRolesController',__name__)


@typesRolesController.route('/typesroles',methods = ['GET'])
async def GetAll():
    try:
        typeRolesServices = typesRolesServices();
        ListTypeRoles = await typeRolesServices.GetAll();
        return json.dumps(ListTypeRoles, default=lambda obj: obj.__dict__), 200;
    except Exception as ex:
        print(ex);
        return jsonify(ex), 500;

@typesRolesController.route('/typesroles/create', methods= ['POST'])
async def Create():
    try:
        data = json.loads(request.data);
        if data['typeName'] != "":
            typeRolesServices = typesRolesServices();
            typeRole = typesRoles(0, data['typeName']);
            result = await typeRolesServices.Save(typeRole);
            return jsonify(result);
        else:
            return jsonify('O tipo de função não pode ser um tipo sem definição(vazio)'), 404;
    except Exception as ex:
        return jsonify(ex), 500;

@typesRolesController.route('/typesroles/delete', methods= ['POST'])
async def Delete():
    try:
        data = json.loads(request.data);
        if data['id'] != "":
            if data['typeName'] != "":
                services = typesRolesServices();
                exists = await services.VerifyTypeRoleNameExists(data['typeName']);
                if exists != False:
                    typerole = await services.GetTypeRoleByTypeName(data['typeName']);
                    if typerole.id == data['id']:
                        res = await services.Delete(typerole.id);
                        return jsonify(res);
                    else:
                        return jsonify("Nenhum registro encontrado, corrija os valores e tente novamente!"), 404;
                else:
                    return jsonify("Nenhum registro encontrado, corrija os valores e tente novamente!"), 404;
            else:
                return jsonify("Não é possivel prosseguir como Tipo de função vazia"), 404;
        else:
            return jsonify("Não é possivel prosseguir com a identificação vazia"), 404;
    except Exception as ex:
        return jsonify(ex), 500;

@typesRolesController.route('/typesroles/getbytypeName', methods= ['POST'])
async def GetBytypeName():
    try:
        data = json.loads(request.data);
        if data['typeName'] != "":
            services = typesRolesServices();
            typerole = await services.GetTypeRoleByTypeName(data['typeName']);
            if typerole != None:
                return jsonify(typerole.__dict__);
            return jsonify("Nada Encontrado"), 404;
        return jsonify("Não é possivel fazer a procura de um dado vazio!"), 404;
    except Exception as ex:
        print(ex)
        return jsonify(ex), 500;
@typesRolesController.route('/typesroles/deletebytypeName', methods = ['POST'])
async def DeleteByTypeName():
    try:
        data = json.loads(request.data);
        if data['typeName'] != "":
            services = typesRolesServices();
            typerole = await services.GetTypeRoleByTypeName(data['typeName']);
            if typerole != None:
                res = await services.Delete(typerole);
                return jsonify(res);
            return jsonify("Tipo de função não encontrado"), 404;
        return jsonify("Não é possivel prosseguir com o campo do tipo de função vazio"), 404;
    except Exception as ex:
        return jsonify(ex), 500;
@typesRolesController.route('/typesroles/deletebyid', methods = ['GET'])
async def DeleteById():
    try:
        data = json.loads(request.data);
        if data['id'] != "":
            services = typesRolesServices();
            result = await services.VerifyTypeRoleExistsById(data['id']);
            if result != False:
                res = await services.Delete(data['id']);
                return jsonify(res);
            else:
                return jsonify("Não existe nenhum dado relacionado a identificação!"), 404;

        else:
            return jsonify("Não é possivel prosseguir com um id vazio!"), 404;
    except Exception as ex:
        return jsonify(ex), 500;

@typesRolesController.route('/typesroles/getByid', methods = ['POST'])
async def GetById():
    try:
        data = json.loads(request.data);
        if data['id'] != "":
            services = typesRolesServices();
            typerole = await services.GetById(data['id']);
            return jsonify(typerole.__dict__);
        else:
            return jsonify("Não é possivel prosseguir com um valor vazio!"), 404;
    except Exception as ex:
        return jsonify(ex), 500;

async def Update():
    return
@typesRolesController.route('/typesroles/teste')
async def Teste():
    services = typesRolesServices();
    typerole = typesRoles(0,'Operacional');
    exists = await services.CheckPropertyTypeRoleExists(typerole);
    return jsonify(exists);
